raft/log_unstable.go

```diff
@@ -85,6 +85,17 @@ func (u *unstable) stableTo(i, t uint64) {
 	if gt == t && i >= u.offset {
 		u.entries = u.entries[i+1-u.offset:]
 		u.offset = i + 1
+		// Discard the underlying array used by the slice if most of it isn't being
+		// used to avoid holding onto a bunch of potentially large entries that
+		// aren't needed anymore. Simply clearing the entries wouldn't be safe
+		// because clients might still be using them.
+		if len(u.entries) == 0 {
+			u.entries = nil
+		} else if len(u.entries) < cap(u.entries)/4 {
```

Comment:
It's pretty arbitrary. I've added a comment explaining the choice, but having a good performance benchmark to justify a constant here would be nice.